src/test/java/com/amazon/ion/impl/macro/MacroCompilerTest.kt

```diff
@@ -0,0 +1,225 @@
+import java.math.BigInteger
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertNull
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+import org.junit.jupiter.api.assertThrows
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.MethodSource
+import org.junit.jupiter.params.provider.ValueSource
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class MacroCompilerTest {
+
+    val ion = IonSystemBuilder.standard().build()
+
+    private data class MacroSourceAndTemplate(val source: String, val template: TemplateMacro) : Arguments {
+        override fun get(): Array<Any> = arrayOf(source, template.signature, template.body)
+    }
+
+    private fun annotations(vararg a: String): List<SymbolToken> = a.map { FakeSymbolToken(it, -1) }
+
+    private infix fun String.shouldCompileTo(macro: TemplateMacro) = MacroSourceAndTemplate(this, macro)
+
+    private fun testCases() = listOf(
+        "(macro empty ())" shouldCompileTo TemplateMacro(emptyList(), emptyList()),
+        "(macro identity (x) x)" shouldCompileTo TemplateMacro(
+            listOf(Parameter("x", Tagged, grouped = false)),
+            listOf(Variable(0)),
+        ),
+        "(macro identity (any::x) x)" shouldCompileTo TemplateMacro(
+            listOf(Parameter("x", Tagged, grouped = false)),
+            listOf(Variable(0)),
+        ),
+        "(macro pi () 3.141592653589793)" shouldCompileTo TemplateMacro(
+            signature = emptyList(),
+            body = listOf(DecimalValue(emptyList(), BigDecimal("3.141592653589793")))
+        ),
+        "(macro group_test ([x]))" shouldCompileTo TemplateMacro(
+            signature = listOf(Parameter("x", Tagged, grouped = true)),
+            body = emptyList()
+        ),
+        // Outer 'values' call allows multiple expressions in the body
+        // The second `values` is a macro call that has a single argument: the variable `x`
+        // The `quote` call causes the third (inner) `(values x)` to be an uninterpreted s-expression.
+        """(macro quote_test (x) (values (values x) (quote (values x))))""" shouldCompileTo TemplateMacro(
+            signature = listOf(Parameter("x", Tagged, grouped = false)),
+            body = listOf(
+                MacroInvocation(ByName("values"), startInclusive = 0, endInclusive = 5),
+                MacroInvocation(ByName("values"), startInclusive = 1, endInclusive = 2),
+                Variable(0),
+                SExpValue(emptyList(), startInclusive = 3, endInclusive = 5),
+                SymbolValue(emptyList(), FakeSymbolToken("values", -1)),
+                SymbolValue(emptyList(), FakeSymbolToken("x", -1)),
+            ),
+        ),
+        "(macro each_type () null true 1 ${"9".repeat(50)} 1e0 1d0 2024-01-16T \"foo\" (quote bar) [] (quote ()) {} {{}} {{\"\"}} )" shouldCompileTo TemplateMacro(
```

Comment:
As mentioned in my comment in the source, this is one that I wasn't sure was allowed.